### ISSUE TITLE
Find the correct start and end position when extract feature

### ISSUE THREAD

**Issue by tiennvcs** [NielsRogge/Transformers-Tutorials#38]
Follow your work, I reproduce the code to train **layoutlmv2** on the **docvqa** dataset. But I have a problem with encoding datasets. Especially, the implementation can't find the extract start and end position of the answer in tokens extracted from the image. 

```python
def subfinder(words_list, answer_list):
    matches = []
    start_indices = []
    end_indices = []
    for idx, i in enumerate(range(len(words_list))):
        if words_list[i] == answer_list[0] and words_list[i:i + len(answer_list)] == answer_list:
            matches.append(answer_list)
            start_indices.append(idx)
            end_indices.append(idx + len(answer_list) - 1)
    if len(matches) != 0:
        return matches[0], start_indices[0], end_indices[0]
    else:
        return None, 0, 0

def encode_dataset(examples, max_length=512):

    images         = [Image.open(image_file).convert("RGB") for image_file in examples['image']]
    org_shapes     = [img.size[0:2] for img in images]

    words          = []
    bbox           = []
    for i in range(len(images)):
        words_img, boxes_img = read_ocr_annotation(file_path=examples['ocr_output_file'][i], shape=org_shapes[i])
        words.append(words_img)
        bbox.append(boxes_img)

    questions  = examples['question']
    encoding   = processor(images, questions, words, bbox, max_length=max_length, padding="max_length", truncation=True)

    # next, add start_positions and end_positions
    start_positions = [0]*BATCH_SIZE
    end_positions   = [0]*BATCH_SIZE

    answers = examples['answers']
    
    # for every example in the batch:
    for idx in range(len(answers)):
        cls_index = encoding.input_ids[idx].index(processor.tokenizer.cls_token_id)

        words_example = [word.lower() for word in words[idx]]

        for answer in answers[idx]:
            match, word_idx_start, word_idx_end = subfinder(words_example, answer.lower().split())
            if match != None:
                break
    
        if match != None:
            sequence_ids = encoding.sequence_ids(idx)
            
            # Start token index of the current span in the text.
            token_start_index = 0
            while sequence_ids[token_start_index] != 1:
                token_start_index += 1

            # End token index of the current span in the text.
            token_end_index = len(encoding.input_ids[idx]) - 1
            while sequence_ids[token_end_index] != 1:
                token_end_index -= 1

            word_ids = encoding.word_ids(idx)[token_start_index:token_end_index+1]
            for id in word_ids:
                if id == word_idx_start:
                    start_positions[idx] = token_start_index
                    break
                else:
                    token_start_index += 1

            for id in word_ids[::-1]:
                if id == word_idx_end:
                    end_positions[idx] = token_end_index
                    break
                else:
                    token_end_index -= 1
        else:
            start_positions[idx] = cls_index
            end_positions[idx] = cls_index


    encoding['start_positions'] = start_positions
    encoding['end_positions']   = end_positions
    encoding['question_id']     = examples['questionId']

    return encoding
```

Could you help with any idea for getting correct annotation as much as possible?

**Comment by sujit420** [comment]
def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]


def read_ocr_results(ocr_json, file_path):
    img = cv2.imread(file_path)
    length, width = img.shape[:2]
    words = []
    boxes = []
    boxes_norm = []
    line_indices = []
    lines_array = []
    for obj in ocr_json["recognitionResults"]:
        lines = obj["lines"]
        idx = 0
        for line in lines:
            lines_array.append(line["text"])
            for word in line["words"]:
                words.append(word["text"].lower())
                line_indices.append(idx)
                x1, y1, x2, y2, x3, y3, x4, y4 = word["boundingBox"]
                new_x1 = min([x1, x2, x3, x4])
                new_x2 = max([x1, x2, x3, x4])
                new_y1 = min([y1, y2, y3, y4])
                new_y2 = max([y1, y2, y3, y4])
                boxes.append([new_x1, new_y1, new_x2, new_y2])
                box_norm = bbox_string([new_x1, new_y1, new_x2, new_y2], width, length)
                assert new_x2 >= new_x1
                assert new_y2 >= new_y1
                assert box_norm[2] >= box_norm[0]
                assert box_norm[3] >= box_norm[1]

                boxes_norm.append(box_norm)
                idx += 1
    return words, boxes_norm


def get_ocr_words_and_boxes(examples):

    # get a batch of document images
    images = [
        Image.open(root_dir + image_file).convert("RGB")
        for image_file in examples["image"]
    ]

    # resize every image to 224x224 + apply tesseract to get words + normalized boxes
    encoded_inputs = feature_extractor(images)

    examples["pixels"] = encoded_inputs.pixel_values
    words = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[0]
        for image_file in examples["image"]
    ]
    boxes = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[1]
        for image_file in examples["image"]
    ]
    examples["words"] = words
    examples["boxes"] = boxes

    return examples


# train_dataset_with_ocr = train_dataset.map(
#     get_ocr_words_and_boxes, batched=True, batch_size=10, num_proc=5
# )

# val_dataset_with_ocr = val_dataset.map(
#     get_ocr_words_and_boxes, batched=True, batch_size=10, num_proc=5
# )
# train_dataset_with_ocr.save_to_disk("train_32k")
# val_dataset_with_ocr.save_to_disk("val_7k")

# print("features saved.....")

# train_dataset_with_ocr = ldv("train_32k")
# val_dataset_with_ocr = ldv("val_7k")


def subfinder(words_list, answer_list):
    matches = []
    start_indices = []
    end_indices = []
    for idx, i in enumerate(range(len(words_list))):
        if len(answer_list) > 0:
            if (
                words_list[i] == answer_list[0]
                and words_list[i : i + len(answer_list)] == answer_list
            ):
                matches.append(answer_list)
                start_indices.append(idx)
                end_indices.append(idx + len(answer_list) - 1)
    if matches:
        return matches[0], start_indices[0], end_indices[0]
    else:
        return None, 0, 0


model_checkpoint = "microsoft/layoutlmv2-base-uncased"

tokenizer = LayoutLMv2TokenizerFast.from_pretrained(model_checkpoint)

# no_answer = 0
# answer_found = 0


def encode_dataset(examples, max_length=512):
    # take a batch
    questions = examples["question"]
    words = examples["words"]
    boxes = examples["boxes"]

    # encode it
    encoding = tokenizer(
        questions,
        words,
        boxes,
        max_length=max_length,
        padding="max_length",
        truncation=True,
    )

    # next, add start_positions and end_positions
    start_positions = []
    end_positions = []
    answers = examples["answers"]
    # for every example in the batch:
    for batch_index in range(len(answers)):
        # print("Batch index:", batch_index)
        cls_index = encoding.input_ids[batch_index].index(tokenizer.cls_token_id)
        # try to find one of the answers in the context, return first match
        words_example = [word.lower() for word in words[batch_index]]
        for answer in answers[batch_index]:
            match, word_idx_start, word_idx_end = subfinder(
                words_example, answer.lower().split()
            )
            if match:
                break
        # EXPERIMENT (to account for when OCR context and answer don't perfectly match):
        if not match:
            for answer in answers[batch_index]:
                for i in range(len(answer)):
                    # drop the ith character from the answer
                    answer_i = answer[:i] + answer[i + 1 :]
                    # check if we can find this one in the context
                    match, word_idx_start, word_idx_end = subfinder(
                        words_example, answer_i.lower().split()
                    )
                    if match:
                        break
        # END OF EXPERIMENT

        if match:
            sequence_ids = encoding.sequence_ids(batch_index)
            # Start token index of the current span in the text.
            token_start_index = 0
            while sequence_ids[token_start_index] != 1:
                token_start_index += 1

            # End token index of the current span in the text.
            token_end_index = len(encoding.input_ids[batch_index]) - 1
            while sequence_ids[token_end_index] != 1:
                token_end_index -= 1

            word_ids = encoding.word_ids(batch_index)[
                token_start_index : token_end_index + 1
            ]
            for id in word_ids:
                if id == word_idx_start:
                    start_positions.append(token_start_index)
                    break
                else:
                    token_start_index += 1

            for id in word_ids[::-1]:
                if id == word_idx_end:
                    end_positions.append(token_end_index)
                    break
                else:
                    token_end_index -= 1

            # print("Verifying start position and end position:")
            # print("True answer:", answer)
            if (
                len(start_positions) == batch_index + 1
                and len(end_positions) == batch_index + 1
            ):
                start_position = start_positions[batch_index]
                end_position = end_positions[batch_index]
                reconstructed_answer = tokenizer.decode(
                    encoding.input_ids[batch_index][start_position : end_position + 1]
                )
                # print("Reconstructed answer:", reconstructed_answer)
                # print("-----------")

            else:
                # print("Answer not found in context in given batch_index")
                # print("-----------")
                if len(start_positions) == batch_index:
                    start_positions.append(cls_index)
                if len(end_positions) == batch_index:
                    end_positions.append(cls_index)

        else:
            # print("Answer not found in context")
            # print("-----------")
            start_positions.append(cls_index)
            end_positions.append(cls_index)

    encoding["image"] = examples["pixels"]
    encoding["start_positions"] = start_positions
    encoding["end_positions"] = end_positions

    return encoding



try to use this and let me know

**Comment by sujit420** [comment]
def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]


def read_ocr_results(ocr_json, file_path):
    img = cv2.imread(file_path)
    length, width = img.shape[:2]
    words = []
    boxes = []
    boxes_norm = []
    line_indices = []
    lines_array = []
    for obj in ocr_json["recognitionResults"]:
        lines = obj["lines"]
        idx = 0
        for line in lines:
            lines_array.append(line["text"])
            for word in line["words"]:
                words.append(word["text"].lower())
                line_indices.append(idx)
                x1, y1, x2, y2, x3, y3, x4, y4 = word["boundingBox"]
                new_x1 = min([x1, x2, x3, x4])
                new_x2 = max([x1, x2, x3, x4])
                new_y1 = min([y1, y2, y3, y4])
                new_y2 = max([y1, y2, y3, y4])
                boxes.append([new_x1, new_y1, new_x2, new_y2])
                box_norm = bbox_string([new_x1, new_y1, new_x2, new_y2], width, length)
                assert new_x2 >= new_x1
                assert new_y2 >= new_y1
                assert box_norm[2] >= box_norm[0]
                assert box_norm[3] >= box_norm[1]

                boxes_norm.append(box_norm)
                idx += 1
    return words, boxes_norm


def get_ocr_words_and_boxes(examples):

    # get a batch of document images
    images = [
        Image.open(root_dir + image_file).convert("RGB")
        for image_file in examples["image"]
    ]

    # resize every image to 224x224 + apply tesseract to get words + normalized boxes
    encoded_inputs = feature_extractor(images)

    examples["pixels"] = encoded_inputs.pixel_values
    words = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[0]
        for image_file in examples["image"]
    ]
    boxes = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[1]
        for image_file in examples["image"]
    ]
    examples["words"] = words
    examples["boxes"] = boxes

    return examples

**Comment by sujit420** [comment]
`def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]


def read_ocr_results(ocr_json, file_path):
    img = cv2.imread(file_path)
    length, width = img.shape[:2]
    words = []
    boxes = []
    boxes_norm = []
    line_indices = []
    lines_array = []
    for obj in ocr_json["recognitionResults"]:
        lines = obj["lines"]
        idx = 0
        for line in lines:
            lines_array.append(line["text"])
            for word in line["words"]:
                words.append(word["text"].lower())
                line_indices.append(idx)
                x1, y1, x2, y2, x3, y3, x4, y4 = word["boundingBox"]
                new_x1 = min([x1, x2, x3, x4])
                new_x2 = max([x1, x2, x3, x4])
                new_y1 = min([y1, y2, y3, y4])
                new_y2 = max([y1, y2, y3, y4])
                boxes.append([new_x1, new_y1, new_x2, new_y2])
                box_norm = bbox_string([new_x1, new_y1, new_x2, new_y2], width, length)
                assert new_x2 >= new_x1
                assert new_y2 >= new_y1
                assert box_norm[2] >= box_norm[0]
                assert box_norm[3] >= box_norm[1]

                boxes_norm.append(box_norm)
                idx += 1
    return words, boxes_norm


def get_ocr_words_and_boxes(examples):

    # get a batch of document images
    images = [
        Image.open(root_dir + image_file).convert("RGB")
        for image_file in examples["image"]
    ]

    # resize every image to 224x224 + apply tesseract to get words + normalized boxes
    encoded_inputs = feature_extractor(images)

    examples["pixels"] = encoded_inputs.pixel_values
    words = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[0]
        for image_file in examples["image"]
    ]
    boxes = [
        read_ocr_results(
            json.load(
                open(ocr_dir + image_file.split("/")[1].replace(".png", ".json"), "r")
            ),
            ocr_dir + "../documents/" + image_file.split("/")[1],
        )[1]
        for image_file in examples["image"]
    ]
    examples["words"] = words
    examples["boxes"] = boxes

    return examples
`

**Comment by sujit420** [comment]
def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]

**Comment by sujit420** [comment]
`def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]`

**Comment by sujit420** [comment]
def bbox_string(box, width, length):
    return [
        int(1000 * (box[0] / width)),
        int(1000 * (box[1] / length)),
        int(1000 * (box[2] / width)),
        int(1000 * (box[3] / length)),
    ]
